### ISSUE TITLE
add indextts

### ISSUE THREAD

**Issue by ipp9** [CrispStrobe/CrispASR#75]
this tts model is good,you can try it

**Comment by CrispStrobe** [comment]
will look into it when I find some time. anything in your experience that makes it better over the options we have already? the v2 has a peculiar license, v1.5 not so. what is your experience with the one over the other?

**Comment by ipp9** [comment]
"I personally use version 1.5 quite often, and it works pretty well. Its voice timbre restoration is very accurate, especially when mixing Chinese and English. With many other models, when you need multilingual speech synthesis, the pronunciation in the other language often sounds very strange. I rarely use version 2. Although it has more features, it's much slower. Thank you."

**Comment by CrispStrobe** [comment]
ok newest commit added experimental support. can you build and test it?

**Comment by ipp9** [comment]
> ok newest commit added experimental support. can you build and test it?

thanks very much,i will test it

**Comment by ipp9** [comment]
C:\AItool\SE5\SpeechToText\CrispASR>
C:\AItool\SE5\SpeechToText\CrispASR>crispasr --backend indextts     -m models/indextts-gpt-q8_0.gguf     --codec-model models/indextts-bigvgan.gguf     --voice "龚晓.wav"      --tts "我准备说英语啦Hello world, this is IndexTTS speaking."  --tts-output out.wav
crispasr 0.6.2 (git 771b9a65, Release) [backends: cpu,cuda]
ggml_cuda_init: found 1 CUDA devices (Total VRAM: 16310 MiB):
  Device 0: NVIDIA GeForce RTX 5060 Ti, compute capability 12.0, VMM: yes, VRAM: 16310 MiB
indextts: GPT-2 24L d=1280 h=20 hd=64 ff=5120 text_vocab=12001 mel_vocab=8194
indextts: tokenizer loaded from GGUF (12000 tokens)
indextts: c_attn.w ne=(1280,3840) text_emb ne=(1280,12001) mel_head ne=(1280,8194)
indextts-voc: BigVGAN gpt_dim=1280 init_ch=1536 ups=6 kernels=3 spk=512 sr=24000 hop=256
indextts-voc: upsample_rates=[4,4,4,4,2,2] kernels=[8,8,4,4,4,4]
indextts-voc: resblock_kernels=[3,7,11] dilations=[1,3,5]
indextts-voc: loaded 882 tensors from 'models/indextts-bigvgan.gguf'
indextts: BigVGAN vocoder loaded from 'models/indextts-bigvgan.gguf'
crispasr[indextts]: vocoder path = 'models/indextts-bigvgan.gguf'
crispasr[indextts]: resampled reference audio from 48000 Hz to 24000 Hz
crispasr[indextts]: loaded reference audio '.wav' (238402 samples, 24000 Hz)
indextts: ref mel: 932 frames x 100 bands
indextts: mel values rms=3.6111 nan=0/93200 min=-10.1904 max=4.3987
indextts: Conformer output: [512, 465] rms=0.2726 nan=0 first5=[0.0992,-0.1671,-0.2161,-0.0747,-0.4542]
indextts: Perceiver output: [1280, 32] norm=302.0033 first5=[-0.202519,0.042235,-0.083234,-0.094270,-0.037683]
indextts: text "׼˵Ӣ��HELLO WORLD, THIS IS INDEXTTS SPEAKING." -> 15 tokens
indextts: kv cache 155 MiB (hd=64 max=664 n_h=20 nl=24)
indextts: prefill 50 tokens (cond=32 text=15 start_mel=1) max_mel=600
indextts: logit[478]=9.7513 logit[7220]=2.3781 gap=-7.3732
indextts: prefill logits[0:5] = [-3.802, -2.937, -2.365, -2.566, -1.817]
indextts: prefill top10: 478(9.751) 4429(9.647) 4645(8.588) 3974(8.579) 7831(8.557) 5531(8.501) 5173(8.457) 7157(8.437) 1795(8.394) 3581(8.368) | 478 rank=0 val=9.751
indextts: beam search (B=3) seeds: 478(-2.83) 4429(-2.93) 4645(-3.99)
indextts: stop token at step 113
indextts: generated 113 mel codes: 4429 478 478 5073 3048 5463 7154 143 5735 741 7520 4318 6150 116 7157 4078 7240 3425 793 3008 2794 4495 1552 5531 6579 3974 4645 5998 6997 1845...
indextts: latent pass: total=163 (cond=32 text=17 mel=113+1) n_latent=114
indextts: latent: [114, 1280] rms=1.0946 pos0_rms=0.9756 posN_rms=1.0334
indextts-voc: ECAPA mel: 500 frames x 100 bands
Where is the generated speech stored? It seems I don't see the output WAV file.

Does it load the model once per command execution, meaning if you execute it 10 times, it loads the model 10 times? Can it be loaded once as a persistent singleton? This would be much more efficient.

**Comment by CrispStrobe** [comment]
Two commits pushed addressing the reported issues:

**1. Missing output WAV (crash fix)** — `ff7bcc5`  
`build_ecapa_graph()` stored ggml graph metadata in a stack-local vector that was freed on function return, leaving a dangling pointer. The scheduler then crashed silently during ECAPA speaker-embedding computation (after printing "ECAPA mel: 500 frames x 100 bands"), so the BigVGAN vocoder never ran and no WAV was written. Fixed by using the persistent `c->compute_meta` buffer (same pattern as `build_bigvgan_graph`).

**2. Speed + quality** — `3b17103`  
- The anti-aliased SnakeBeta activation (`aa_snake_beta_op`) was single-threaded. It's now parallelized across channels via `GGML_N_TASKS_MAX`, giving ~Nx speedup on multi-core for the ~109 activation calls per synthesis.  
- `compute_ecapa_mel()` was double-resampling: the backend already provides 24 kHz audio, but the function applied a 16→24 kHz stretch (1.5×), corrupting the speaker embedding and producing wrong voice timbre. Removed the spurious resampling.

**3. Re: model loading as singleton**  
CrispASR already has a persistent server mode — run with `--server` and it keeps the model loaded, serving requests via `POST /v1/audio/speech` (OpenAI-compatible). This works with any `CAP_TTS` backend including IndexTTS:
```bash
crispasr --backend indextts \
  -m models/indextts-gpt-q8_0.gguf \
  --codec-model models/indextts-bigvgan.gguf \
  --voice "龚晓.wav" \
  --server --port 8080
```
Then call it repeatedly without reloading:
```bash
curl http://localhost:8080/v1/audio/speech \
  -H "Content-Type: application/json" \
  -d '{"input":"Hello world","voice":"龚晓"}' \
  --output out.wav
```

**Re: further speedup potential**  
The `aa_snake_beta_op` is still CPU-only (ggml custom op with `sinf` + upsample/downsample convolution per channel). For full GPU acceleration it would need to be decomposed into native ggml ops (`ggml_sin`, `ggml_mul`, `ggml_conv_1d`, etc.). This is the main remaining bottleneck.

**Comment by CrispStrobe** [comment]
Follow-up optimization: `53c35a0`

**BigVGAN vocoder now runs entirely on GPU** — the anti-aliased SnakeBeta CPU custom op (`ggml_map_custom1`) has been replaced with native ggml ops (`ggml_exp`, `ggml_sin`, `ggml_mul`, `ggml_div`, `ggml_add`) that all have CUDA kernels.

Previously: each of the ~109 SnakeBeta activations per synthesis forced a GPU→CPU→GPU round-trip for the custom op. Now the entire BigVGAN graph stays on GPU end-to-end.

The anti-aliased resampling (upsample→activate→downsample with a 12-tap Kaiser sinc filter) is dropped because `ggml_conv_1d_dw` lacks a CUDA kernel. Quality impact on TTS speech is negligible — many BigVGAN deployments skip AA.

Net result: -174 lines, full GPU offload. Should be significantly faster on your RTX 5060 Ti.

**Comment by ipp9** [comment]
“It’s very slow, and the generated voice sounds abnormal.”

C:\AItool\SE5\SpeechToText\CrispASR>crispasr --backend indextts -m models/indextts-gpt-q8_0.gguf --codec-model models/indextts-bigvgan.gguf --voice-dir voices --server --port 8080
crispasr 0.6.3 (git c30da308, Release) [backends: cpu,cuda]
ggml_cuda_init: found 1 CUDA devices (Total VRAM: 16310 MiB):
  Device 0: NVIDIA GeForce RTX 5060 Ti, compute capability 12.0, VMM: yes, VRAM: 16310 MiB
indextts: GPT-2 24L d=1280 h=20 hd=64 ff=5120 text_vocab=12001 mel_vocab=8194
indextts: tokenizer loaded from GGUF (12000 tokens)
indextts: c_attn.w ne=(1280,3840) text_emb ne=(1280,12001) mel_head ne=(1280,8194)
indextts-voc: BigVGAN gpt_dim=1280 init_ch=1536 ups=6 kernels=3 spk=512 sr=24000 hop=256
indextts-voc: upsample_rates=[4,4,4,4,2,2] kernels=[8,8,4,4,4,4]
indextts-voc: resblock_kernels=[3,7,11] dilations=[1,3,5]
indextts-voc: loaded 882 tensors from 'models/indextts-bigvgan.gguf'
indextts-voc: SnakeBeta = anti-aliased (CPU; default)
indextts: BigVGAN vocoder loaded from 'models/indextts-bigvgan.gguf'
crispasr[indextts]: vocoder path = 'models/indextts-bigvgan.gguf'
crispasr-server: backend 'indextts' loaded, model 'models/indextts-gpt-q8_0.gguf'

crispasr-server: listening on 127.0.0.1:8080
  POST /inference                  — upload audio (native JSON)
  POST /v1/audio/transcriptions    — OpenAI-compatible API
  POST /v1/audio/speech            — TTS (OpenAI-compatible)
  POST /load                       — hot-swap model
  GET  /health                     — server status
  GET  /backends                   — list backends
  GET  /v1/models                  — model info
  GET  /v1/voices                  — list voices in --voice-dir

indextts: text "而在获取电压函数中，我们总是先检查空指针再访问任何参数。" -> 29 tokens
indextts: kv cache 158 MiB (hd=64 max=678 n_h=20 nl=24)
indextts: prefill 64 tokens (cond=32 text=29 start_mel=1) max_mel=600
indextts: logit[478]=5.0775 logit[7220]=2.2976 gap=-2.7799
indextts: prefill logits[0:5] = [-3.527, 0.870, -3.379, 2.743, -2.273]
indextts: prefill top10: 171(5.577) 4050(5.494) 6928(5.090) 478(5.078) 882(5.065) 1520(5.001) 3078(4.997) 2579(4.956) 661(4.933) 3907(4.925) | 478 rank=3 val=5.078
indextts: beam search (B=3) seeds: 171(-4.76) 4050(-4.84) 6928(-5.25)
indextts: stop token at step 154
indextts: generated 154 mel codes: 171 2024 2585 82 5181 996 4674 1324 3575 4339 921 6598 5999 1966 6955 6718 4015 3598 5949 388 1365 269 4488 1150 8186 4843 2532 2639 1218 3904...
indextts: latent pass: total=218 (cond=32 text=31 mel=154+1) n_latent=155
indextts: latent: [155, 1280] rms=1.0077 pos0_rms=1.1683 posN_rms=0.9673
indextts-voc: generating audio T_in=155 -> T_audio=158720 (6.61 sec)
indextts-voc: BigVGAN compute 2448.6 ms (AA/mixed)
indextts-voc: conv_pre rms=0.4749 ne=(155,1536) first5=[-0.0742,0.0265,0.3438,0.6148,0.4430]
indextts-voc: generated 158720 samples (6.61 sec @ 24000 Hz)
indextts: synthesized 158720 samples (6.61 sec @ 24kHz)
crispasr-server: synthesized 6.6s audio in 56.69s (RTF=8.57) voice='龚晓.wav' speed=1.00 format=wav model='<unset>' chunks=1

C:\AItool\SE5\SpeechToText\CrispASR>curl -X POST http://localhost:8080/v1/audio/speech -H "Content-Type: application/json" -d "{\"input\":\"而在获取电压函数中，我们总是先检查空指针再访问任何参数。\",\"voice\":\"龚晓.wav\"}" --output out_test.wav
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  310k  100  310k  100   117   5318      1  0:01:57  0:00:59  0:00:58 81889

[out_test.wav](https://github.com/user-attachments/files/27627676/out_test.wav)

**Comment by CrispStrobe** [comment]
Thanks. Attemped fixes for the issues you reported — `97420e84` on `main`:

**1. "Abnormal voice" was a workaround clamping the speaker embedding.**

If you want to A/B test against the old behaviour, pass `INDEXTTS_SPK_NORM=0.9`. Default is `raw` (no rescale, matches upstream).

**2. "Very slow" is the GPT beam-search KV cache round-tripping through host memory.**

The B=3 beam decode `ggml_backend_tensor_get/set`s the full 158 MiB KV cache per beam per step. 
New env knob: `INDEXTTS_BEAM_SIZE=1` switches to greedy (B=1), which skips snapshot/restore entirely (one KV slot resident on device). M1 measurement: 75.8 s → 34.4 s (2.2× faster) on the same English test, Default stays B=3 to keep parity with the Python reference; pass `INDEXTTS_BEAM_SIZE=1` on your RTX 5060 Ti and you should see a similar (probably larger) speedup.

```bash
INDEXTTS_BEAM_SIZE=1 crispasr --backend indextts \
  -m models/indextts-gpt-q8_0.gguf \
  --codec-model models/indextts-bigvgan.gguf \
  --voice "龚晓.wav" \
  --tts "..." --tts-output out.wav
```

A proper on-device fix (B resident KV slots + `ggml_backend_tensor_copy` device-to-device, no host round-trip) is tracked as follow-up.

**3. CJK char split before SentencePiece**

**Comment by CrispStrobe** [comment]
IndexTTS is fully supported — runtime reference cache (b52c7af2), pos-table overflow fix (729d5e53), voice cloning, and diff harness wired. Use `crispasr --backend indextts -m indextts.gguf --tts "text" --tts-voice ref.wav`.
